### ISSUE TITLE
[assignment-A2-mlp] DenseMLPWithLoRA实现求助 测试失败 - 即使在 LoRA rank=0 的情况下也不匹配

### ISSUE THREAD

**Issue by xrlexpert** [Strivin0311/llms-learning#4]
## 问题描述
我严格按照任务描述实现了 DenseMLPWithLoRA 类，但测试仍然失败。更令人困惑的是，即使在 LoRA rank=0 的情况下（此时应该只使用基础MLP而不涉及LoRA），我的实现与参考实现仍然有显著差异。
## 测试样例
```
"task1": {
        "case1": {
            "b": 1,
            "s": 1,
            "h": 4,
            "ffh": 8,
            
            "activation_type": MLPActivationTypeRef.SIGMOID,
            
            "r": 0,
            "alpha": None,
            "dropout": 0.0,
            
            "activation_dtype": torch.bfloat16,
            "activation_device": "cpu",
        }
    },
```
测试输出显示，在 LoRA rank=0 的情况下：
```
input: tensor([[[0.3359, 0.1289, 0.2344, 0.2305]]], dtype=torch.bfloat16)
r: 0
output: tensor([[[-0.1494, -0.2256, -0.0708, -0.0986]]], grad_fn=<ToCopyBackward0>)
output_ref: tensor([[[-0.1807, -0.0295, -0.0903,  0.0310]]], grad_fn=<ToCopyBackward0>)
```
## 我的实现
``` python
from enum import Enum
from typing import Optional

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.distributed import ProcessGroup


class MLPActivationType(Enum):
    RELU = "relu"
    GELU = "gelu"
    SILU = "silu"
    SIGMOID = "sigmoid"
    BILINEAR = "bilinear"

# Class-level activation function mapping
ACT_FN_MAP = {
    MLPActivationType.RELU: F.relu,
    MLPActivationType.GELU: F.gelu,
    MLPActivationType.SILU: F.silu,
    MLPActivationType.SIGMOID: F.sigmoid,
    MLPActivationType.BILINEAR: lambda x: x * x,  # or other suitable implementation
}

class DenseMLPWithLoRA(nn.Module):
    """Dense MLP module with LoRA adapters
    This is a GLU-style dense MLP layer with LoRA adapters.
    """
    
    def __init__(self,
        hidden_size: int,
        ffh_size: int,
        activation_type: MLPActivationType = MLPActivationType.SILU,
        init_base_seed: int = 42,
        lora_rank: int = 0,
        lora_alpha: Optional[float] = None,
        lora_dropout_rate: float = 0.0,
        lora_dropout_seed: int = 42,
        lora_init_base_seed: int = 42,
        dtype: torch.dtype = torch.float32,
        device: str = "cpu",
    ):
        """Initialize Dense MLP module with LoRA adapters
        Args:
            hidden_size(int): hidden dimension size
            ffh_size(int): hidden dimension size
            activation_type(MLPActivationType, default = "silu"): activation type
            init_base_seed(int, default = 42): seed for base weight initialization
            lora_rank(int, default = 0): lora rank, if 0, then no lora to apply
            lora_alpha(Optional[float], default = None): lora alpha, if None, then set to lora_rank
            lora_dropout_rate(float, default = 0.0): lora dropout rate
            lora_dropout_seed(int, default = 42): lora dropout seed
            lora_init_base_seed(int, default = 42): seed for lora weight initialization
            dtype(torch.dtype, default = torch.float32): parameter dtype
            device(str, default = "cpu"): parameter device
        """
        super().__init__()
        self.hidden_size = hidden_size
        self.ffh_size = ffh_size
        self.activation_type = activation_type
        self.init_base_seed = init_base_seed
        self.lora_rank = lora_rank
        self.lora_alpha = lora_rank if lora_alpha is None else lora_alpha
        self.lora_dropout_rate = lora_dropout_rate
        self.lora_dropout_seed = lora_dropout_seed
        self.lora_init_base_seed = lora_init_base_seed
        self.dtype = dtype
        self.device = device

        self.act_fn = ACT_FN_MAP[self.activation_type]
        
        # Create linear layers without specifying bias
        self.gate_proj = nn.Linear(self.hidden_size, self.ffh_size, bias=False)
        self.up_proj = nn.Linear(self.hidden_size, self.ffh_size, bias=False)
        self.down_proj = nn.Linear(self.ffh_size, self.hidden_size, bias=False)
        
        # Only create LoRA layers when lora_rank > 0
        if self.lora_rank > 0:
            self.lora_a = nn.Linear(self.hidden_size, self.lora_rank, bias=False)
            self.lora_b = nn.Linear(self.lora_rank, self.hidden_size, bias=False)
            self.lora_dropout = nn.Dropout(self.lora_dropout_rate)
        
        # Convert all parameters to the specified dtype and device
        self.to(dtype=self.dtype, device=self.device)
        self.reset_parameters()
    
    def forward(self, input: torch.Tensor) -> torch.Tensor:
        """The forward pass of the Dense MLP module with LoRA adapters
        
        Args:
            input(torch.Tensor): input tensor, with shape: [batch_size, seq_len, hidden_size]
            
        Returns:
            output(torch.Tensor): output tensor, with shape: [batch_size, seq_len, hidden_size]
        """
        # Save the dtype and device of the input
        input_dtype = input.dtype
        input_device = input.device
        
        # Convert the input to the model's dtype and device
        input = input.to(dtype=self.dtype, device=self.device)
        
        # GLU-style MLP forward pass
        gate_output = self.act_fn(self.gate_proj(input))
        up_output = self.up_proj(input)
        print(gate_output.shape, up_output.shape)
        output = self.down_proj(gate_output * up_output)
        
        # Apply LoRA if rank > 0
        if self.lora_rank > 0:
            # LoRA forward: (α/r * x * A * B)
            lora_output = self.lora_a(input)
            lora_output = self.lora_b(lora_output)
            
            # Apply scaling factor α/r
            lora_output = lora_output * (self.lora_alpha / self.lora_rank)
            
            # Apply dropout to LoRA output
            if self.training and self.lora_dropout_rate > 0:
                torch.manual_seed(self.lora_dropout_seed)
                lora_output = self.lora_dropout(lora_output)
            
            # Add LoRA output to the original output
            output = output + lora_output
            
        # Convert the output back to the input's dtype and device
        return output.to(dtype=input_dtype, device=input_device)
    
    def reset_parameters(self):
        """Initialize the weights of the Dense MLP module with LoRA adapters
        from a normal distribution (or a uniform distribution for LoRA weights)
        """
        # Create generator for base MLP weights
        generator = torch.Generator(device=self.device)
        
        # Select initialization method based on activation function type
        if self.activation_type == MLPActivationType.SIGMOID or self.activation_type == MLPActivationType.BILINEAR:
            # Xavier Initialization for sigmoid and bilinear
            generator.manual_seed(self.init_base_seed + 1)
            nn.init.xavier_normal_(self.up_proj.weight, generator=generator)
            
            generator.manual_seed(self.init_base_seed + 2)
            nn.init.xavier_normal_(self.gate_proj.weight, generator=generator)
            
            generator.manual_seed(self.init_base_seed + 3)
            nn.init.xavier_normal_(self.down_proj.weight, generator=generator)
        else:
            # Kaiming Initialization for ReLU family
            generator.manual_seed(self.init_base_seed + 1)
            nn.init.kaiming_normal_(self.up_proj.weight, mode="fan_in", nonlinearity="relu", generator=generator)
            
            generator.manual_seed(self.init_base_seed + 2)
            nn.init.kaiming_normal_(self.gate_proj.weight, mode="fan_in", nonlinearity="relu", generator=generator)
            
            generator.manual_seed(self.init_base_seed + 3)
            nn.init.kaiming_normal_(self.down_proj.weight, mode="fan_in", nonlinearity="relu", generator=generator)
        
        # Initialize LoRA weights (if rank > 0)
        if self.lora_rank > 0:
            # Create generator for LoRA weights
            lora_generator = torch.Generator(device=self.device)
            # Xavier Uniform Initialization for LoRA
            lora_generator.manual_seed(self.lora_init_base_seed + 1)
            nn.init.xavier_uniform_(self.lora_a.weight, generator=lora_generator)
            
            lora_generator.manual_seed(self.lora_init_base_seed + 2)
            nn.init.xavier_uniform_(self.lora_b.weight, generator=lora_generator)
         

```
* 所有线性层不需要bias
* `SIGMOID` 和 `BILINEAR` 采用`xavier_normal_`初始化， 其余采用`kaiming_normal_`初始化，该样例输入为`SIGMOID`
* 按照`up` , `grate`, `down` 顺序依次将种子递增

**Comment by Olynx7** [comment]
gate_proj up_proj down_proj 以及lora用nn.Parameter(torch.empty(size=(hidden_size, ffh_size), dtype=dtype, device=device))，不要用linear，因为两者用xavier normal生成的结果和ref不一样

**Comment by Strivin0311** [comment]
> gate_proj up_proj down_proj 以及lora用nn.Parameter(torch.empty(size=(hidden_size, ffh_size), dtype=dtype, device=device))，不要用linear，因为两者用xavier normal生成的结果和ref不一样

是的，`nn.Linear`实际上存的是W^T

**Comment by xrlexpert** [comment]
您好，修改为nn.Parameter之后我修改ffh参数，只能通过部分样例。另外关于xavier_normal_和kaiming_normal_，我理解的是两个初始化函数默认以传入权重是转置的形式，而nn.Linear存的是W^T,因此调用初始化时就应该是nn.init.xavier_normal_(self.up_proj.weight, generator=generator)而非nn.init.xavier_normal_(self.up_proj.weight.T, generator=generator)

**Comment by xrlexpert** [comment]
我怀疑可能我的实现都是使用input@W的形式计算，与ref不同（例如input@W.T）导致的这个问题吗

**Comment by xrlexpert** [comment]
> gate_proj up_proj down_proj 以及lora用nn.Parameter(torch.empty(size=(hidden_size, ffh_size), dtype=dtype, device=device))，不要用linear，因为两者用xavier normal生成的结果和ref不一样

请问是 gate_proj up_proj down_proj 都用size=(hidden_size, ffh_size)来初始化然后根据需要，选择@W或者@W^T吗
